src/main/java/com/es/phoneshop/security/impl/DefaultDosProtectionService.java

```diff
@@ -0,0 +1,45 @@
+package com.es.phoneshop.security.impl;
+
+import com.es.phoneshop.security.DosProtectionService;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class DefaultDosProtectionService implements DosProtectionService {
+    private static DefaultDosProtectionService instance;
+    private static final long THRESHOLD = 20;
+    private static final long TIME = 60000;
+    private Map<String, Long> countMap = new ConcurrentHashMap<>();
+    private long previousTime;
+
+    private DefaultDosProtectionService() {
+        previousTime = System.currentTimeMillis();
+    }
+
+    public static synchronized DefaultDosProtectionService getInstance() {
+        if (instance == null) {
+            instance = new DefaultDosProtectionService();
+        }
+        return instance;
+    }
+
+    @Override
+    public boolean isAllowed(String ip) {
+        long currentTime = System.currentTimeMillis();
+        if (currentTime - previousTime > TIME) {
+            countMap.clear();
+            previousTime = currentTime;
+        }
```

Comment:
Just to note - multiple threads might access `previousTime` at the same time. So it's possible here to have race conditions situations, though it will work here, since `currentTime` will not be much different for both threads.

Reply:
I would also move this part to separate private method for clarity